### PR TITLE
feature: Added checking for miss flow entries

### LINKED ISSUE #15: Avoid deleting redundant miss flow entries
This [code](https://github.com/kytos-ng/of_multi_table/blob/fe53d597bbac6b7153fe9922f2d20f8a29970957/main.py#L174-L182) deletes miss flow entries every pipeline change. It would be better to avoid deleting every miss flow entry, instead just the necessary ones.

### PR DESCRIPTION
**Pull request by Alopalao**
Closes #15 

### Summary

Improve transition from active to active pipeline, miss flow entries are not longer unnecessarily deleted.
[PR](https://github.com/kytos-ng/flow_manager/pull/156) from `flow_manage` is necessary.

### Local tests
First enable a pipeline:
```
{"multi_table": [
        {
            "table_id": 0,
            "table_miss_flow": {"priority": 0, "instructions": [{"instruction_type": "goto_table", "table_id": 1}]}
        },
        {
            "table_id": 1,
            "napps_table_groups": {"coloring": ["base"]},
            "table_miss_flow": {"priority": 0, "instructions": [{"instruction_type": "goto_table", "table_id": 2}]}
        },
        {
            "table_id": 2,
            "napps_table_groups": {"of_lldp": ["base"]},
            "table_miss_flow": {"priority": 0, "instructions": [{"instruction_type": "goto_table", "table_id": 3}]}
        },
        {
            "table_id": 3,
            "napps_table_groups": {"mef_eline": ["evpl"]},
            "table_miss_flow": {"priority": 0,"instructions": [{"instruction_type": "goto_table","table_id": 4}]}
        },
        {
            "table_id": 4,
            "napps_table_groups": {"mef_eline": ["epl"]}
        }
]}
```

Enable other pipeline:
```
{"multi_table": [
                {
                    "table_id": 1,
                    "napps_table_groups": {"mef_eline": ["epl", "evpl"], "of_lldp": ["base"], "coloring": ["base"]},
                    "table_miss_flow": {"priority": 0, "instructions": [{"instruction_type": "goto_table", "table_id": 2}]}
                },
                {
                    "table_id": 2,
                    "table_miss_flow": {"priority": 0, "instructions": [{"instruction_type": "goto_table", "table_id": 5}]}
                },
                {
                    "table_id": 5,
                    "table_miss_flow": {"priority": 99, "instructions": [{}]}
                }
]}
```
Miss flow entries from:
- Table 0 and 3 are delete
- Table 5 is installed
- Table 2 is modified (deleted and installed)
- Table 1 is not touched

No consistency check was required

### End-to-End Tests

```
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-1.0.0
rootdir: /tests
plugins: timeout-2.1.0, rerunfailures-10.2, anyio-3.6.2
collected 5 items

tests/test_e2e_60_of_multi_table.py .....                                [100%]
```

### CODE REVIEW

File: main.py
Comment (viniarck) [inline main.py]: Can you also add the returned type `-> tuple[dict[str, dict], set[str]]` for completeness here?

Also, could you rename `flow_ids` to be `stored_table_ids` or something else similar that makes it clear it's a table id? `flow_ids` is storing table ids, when I was reading the code the variable name can lead to initial confusion, especially since our flow data models also have an id.

File: main.py
Comment (viniarck) [inline main.py]: You can simplify `ids` type here, since you're just iterating over `ids` and don't necessarily care about the container type, you can use a `Iterable[int]` from `typing`. 

Same comment for `def delete_miss_flows` above

File: main.py
Comment (viniarck) [inline main.py]: Same as my prior comment regarding `flow_ids` if you could rename it more closely related to `stored_table_ids` or something else that would improve understanding of what it's actually holding.

File: tests/unit/test_main.py
Comment (viniarck) [inline tests/unit/test_main.py:276]: Excellent that this cover both installs and deletions, could you also cover another case here when no changes (or additional calls to `mock_install` or `mock_delete`) are expected? I'm not sure if this has been indirectly covered though.

**Review by viniarck — COMMENTED** [review]
LGTM. Nicely done @Alopalao. 

I left a few minor suggestions for testing and variable names. 

I'll approve it once e2e are also passing since this is a relatively significant change for table misses at least :)

File: tests/unit/test_main.py
Comment (Alopalao) [reply tests/unit/test_main.py:276]: It was not. I divided all tests for method into different methods since it was getting too large.

File: main.py
Comment (Alopalao) [reply main.py]: Changed naming.

File: main.py
Comment (Alopalao) [reply main.py]: Was looking for something similar. Added `Iterable[int]` although `Union` was part of a different approach, this method can handle all of those just fine.

File: main.py
Comment (Alopalao) [reply main.py]: Added the return type and change the naming.

**Comment by Alopalao** [comment]
E2E test results after last commit [bcaa302c0b9f2bdb4165a1df55076a3fd91f2bb1](https://github.com/kytos-ng/of_multi_table/pull/22/commits/bcaa302c0b9f2bdb4165a1df55076a3fd91f2bb1):

```
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-1.0.0
rootdir: /tests
plugins: timeout-2.1.0, rerunfailures-10.2, anyio-3.6.2
collected 5 items

tests/test_e2e_60_of_multi_table.py .....                                [100%]
```
[PR](https://github.com/kytos-ng/flow_manager/pull/156) from `flow_manage` is necessary.

File: main.py
Comment (viniarck) [inline main.py]: @Alopalao this looks like a non intentional change, can you double check if you meant `if not flows_by_swich`?

**Review by viniarck — COMMENTED** [review]
Awesome to have the extra unit test cov and e2e test passing. I was going to approve, but just notice another change that I raised in the comments.

File: main.py
Comment (Alopalao) [reply main.py]: In case where `coloring`, `of_lldp` or `telemetry_int` are not installed. At the start of kytos, there are probably no flows installed but a pipeline with miss flows is enabled. The miss flows only need switches to be installed.

Comment (viniarck) [reply main.py]: @Alopalao, this seems like it can lead to a potential issue. Let's verify by parts:

* `coloring`, `of_lldp`, `mef_eline` and future `telemetry_int` are expected to be dependencies and loaded first than `of_multi_table`. We'll structure the docker image (for prod usage, and in development they should follow the same ordering) and recommend it accordingly. As you mentioned, if any flows miss might (or might not) be there, so maybe we don't even need a early return here at all since your diff algo will do what's expected? 

    * Otherwise it might discard an important execution of FlowMods of flow misses (even if the switches might not be connected since we're leveraging `force=true`). Also if a switch has ever existed and it if flows were installed, `topology` will always guarantee that this switch will be loaded first since `topology` has a higher priority in our docker file since most NApps depends on it, and currently we don't support switch hard deletions, so a concern of potentially sending to a non existing switch isn't an issue as far as I can follow.

* If any of these NApps aren't installed, is a valid concern for this iteration, but shouldn't this even be treated way earlier? For instance, before even trying [to enable the pipeline on load_pipeline?](https://github.com/kytos-ng/of_multi_table/blob/master/main.py#L76-L81) since at that point `handle_enable_table` would be stuck since a reply might never come? As it currently is, seems fair game to me though, since eventually these NApps that are missing can also get loaded, maybe a warning message on `load_pipeline` method should suffice if any of the subscribed NApps aren't loaded? and then we can keep `load_pipeline` pretty much as it is but with with an exception warning message if that ever happens?

Comment (viniarck) [reply main.py]: Let me know if you're seeing something else differently or if you have other suggestions

Comment (Alopalao) [reply main.py]: Oh ye, I can just remove it. But for completeness this condition was just for a very specific case:
- Only `mef_eline` is installed/enabled.
- There are no EVCs deployed, therefore no flows.
- A pipeline was enabled and it has miss flow entries in it.

In this case the request GET from `http://localhost:8181/api/kytos/flow_manager/v2/stored_flows?state=installed` is going to return an empty dictionary and the previous condition was going to do an early return without installing the miss flows.

Comment (Alopalao) [reply main.py]: E2E tests results without the condition
```
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-1.0.0
rootdir: /tests
plugins: timeout-2.1.0, rerunfailures-10.2, anyio-3.6.2
collected 5 items

tests/test_e2e_60_of_multi_table.py .....                                [100%]
```

Comment (viniarck) [reply main.py]: Got it, @Alopalao. 

As it is, it's fair game to me. It's great that you also considered this case when one of the pre-requisites NApps weren't there, let's leave it to another future opportunity if it's ever needed and refined. I'll leave it up to you if you think `load_pipeline` should also give a warning or not (if a NApp that was found isn't enabled or subscribed), I'm indifferent, all things considered, especially since in prod usage we know that the pre-requisites NApps that this NApp requires will be enabled.

**Review by viniarck — APPROVED** [review]
Nicely done @Alopalao, great optimization and discussions on this PR. 

Feel free to merge this PR.

**Pull request merged**
